### ISSUE TITLE
[MQTT] Homie V3 device doesn't work properly

### ISSUE THREAD

**Issue by bodiroga** [eclipse/smarthome#6355]
Hi @davidgraeff!

Many thanks for all the work you have done around MQTT, this is a great step forward for Eclipse Smarthome (and openHAB) :+1: 

I've been doing some tests simulating a Homie device with a python script and the results haven't been the desired ones. Here you have the script:

```python
import random
import time
import paho.mqtt.client as mqtt

mqtt_broker_host = "localhost"
mqtt_broker_port = 1883
mqtt_client_id = "homie_3.0_temperature_sensor"
mqtt_client_keepalive = 60

homie_prefix = "homie"
homie_version = "3.0.0"
homie_device_id = "homie-3-temperature-sensor"
homie_device_name = "Homie 3.0 Temperature sensor"
homie_device_ip = "192.168.1.150"
homie_device_mac = "DE:AD:BE:EF:FE:ED"
homie_device_nodes = ["temperature"]
homie_device_state = None

def publish_homie_version(client):
    topic = "{}/{}/{}".format(homie_prefix, homie_device_id, "$homie")
    client.publish(topic, homie_version, 1, True)

def advertize_device():
    client.publish("homie/homie-3-temperature-sensor/$state", "init", 1, True)
    publish_device_info()
    publish_node_info()
    client.publish("homie/homie-3-temperature-sensor/$state", "ready", 1, True)

def publish_device_info():
    client.publish("homie/homie-3-temperature-sensor/$homie", homie_version, 1, True)
    client.publish("homie/homie-3-temperature-sensor/$name", homie_device_name, 1, True)
    client.publish("homie/homie-3-temperature-sensor/$localip", homie_device_ip, 1, True)
    client.publish("homie/homie-3-temperature-sensor/$mac", homie_device_mac, 1, True)
    client.publish("homie/homie-3-temperature-sensor/$fw/name", "sensor-firmware", 1, True)
    client.publish("homie/homie-3-temperature-sensor/$fw/version", "1.0.0", 1, True)
    client.publish("homie/homie-3-temperature-sensor/$nodes", "temperature", 1, True)
    client.publish("homie/homie-3-temperature-sensor/$implementation", "homie-python", 1, True)
    client.publish("homie/homie-3-temperature-sensor/$stats/interval", "60", 1, True)

def publish_node_info():
    client.publish("homie/homie-3-temperature-sensor/temperature/$name", "Indoor temperature", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/$type", "Temperature", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/$properties", "temperature", 1, True)

    client.publish("homie/homie-3-temperature-sensor/temperature/temperatura", 23, 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/temperatura/$name", "Indoor temperature", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/temperatura/$settable", "false", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/temperatura/$unit", "ºC", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/temperatura/$datatype", "float", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/temperatura/$format", "-20:120", 1, True)

def on_connect(client, userdata, flags, rc):
    print("Connected with result code {}".format(rc))
    advertize_device()


def on_message(client, userdata, msg):
    print("{} -> {}".format(msg.topic, msg.payload))


def loop():
    last_temperature_time = 0
    while True:
        if (time.time() - last_temperature_time) > 5:
            temperature_value = random.uniform(20,25)
            print("Sending tempertaura value: {0:.2f}ºC".format(temperature_value))
            client.publish("homie/homie-3-temperature-sensor/temperature/temperatura", "{0:.2f}".format(temperature_value), 1, True)
            last_temperature_time = time.time()


client = mqtt.Client(mqtt_client_id)
client.on_connect = on_connect
client.on_message = on_message

client.will_set("homie/homie-3-temperature-sensor/$state", "lost", 1, True)

client.connect(mqtt_broker_host, mqtt_broker_port, mqtt_client_keepalive)

client.loop_start()

try:
    loop()
except KeyboardInterrupt:
    print("Stopping client")
    client.publish("homie/homie-3-temperature-sensor/$state", "disconnected", 1, True)
    client.disconnect()
    client.loop_stop()
```
Here you have what I tried:

- I have created a brokerConnection to my Mosquitto installation through the inbox (the broker was perfectly discovered)
- The script was executed and all the topics were correctly published.
- The Homie device appeared in the inbox, but after importing it, the thing remained "Offline" (even if the message published in the "$state" attribute is 'ready'). In the "Edit" screen of the thing the "Bridge selection" field was empty and I had to manually set it to "MQTT Broker" (the broker previously discovered).
- The Thing went "Online", but no channel was discovered (nor Thing attribute) and the following message was logged in the openhab.log file:

```
2018-10-12 20:16:12.320 [WARN ] [ternal.mapping.SubtopicFieldObserver] - Tried to add new property temperature, but failed
java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: Label for a ChannelType must not be empty.
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniRun(CompletableFuture.java:708) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:687) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[?:?]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962) ~[?:?]
	at org.eclipse.smarthome.binding.mqtt.generic.internal.mapping.FieldMqttMessageSubscriber.timeoutReached(FieldMqttMessageSubscriber.java:142) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.lang.IllegalArgumentException: Label for a ChannelType must not be empty.
	at org.eclipse.smarthome.core.thing.internal.type.AbstractChannelTypeBuilder.<init>(AbstractChannelTypeBuilder.java:52) ~[?:?]
	at org.eclipse.smarthome.core.thing.internal.type.StateChannelTypeBuilderImpl.<init>(StateChannelTypeBuilderImpl.java:40) ~[?:?]
	at org.eclipse.smarthome.core.thing.type.ChannelTypeBuilder.state(ChannelTypeBuilder.java:94) ~[?:?]
	at org.eclipse.smarthome.binding.mqtt.generic.internal.convention.homie300.Property.attributesReceived(Property.java:155) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniRun(CompletableFuture.java:705) ~[?:?]
	... 11 more
```

Can you try the provided script and confirm if you see the same behavior? The script requires the paho-mqtt library, but you can install it very easily with "pip install paho-mqtt".

Thanks again for your help and support and looking forward to hearing from you :wink: 

Best regards,

Aitor

PS: I can provide a video showing a step by step of the process, just ask for it :+1:  How can I change the log level of the binding to debug? I have tried "log:set DEBUG org.eclipse.smarthome.binding.mqtt.generic" with no result :cry:

**Comment by bodiroga** [comment]
Ouch! I have discovered an error in the script and now the channel is correctly rendered and updated :man_facepalming: 

The property was named "temperatura", but it was announced as "temperature". Here you have the corrected version of the script:

```python
import random
import time
import paho.mqtt.client as mqtt

mqtt_broker_host = "localhost"
mqtt_broker_port = 1883
mqtt_client_id = "homie_3.0_temperature_sensor"
mqtt_client_keepalive = 60

homie_prefix = "homie"
homie_version = "3.0.0"
homie_device_id = "homie-3-temperature-sensor"
homie_device_name = "Homie 3.0 Temperature sensor"
homie_device_ip = "192.168.1.150"
homie_device_mac = "DE:AD:BE:EF:FE:ED"
homie_device_nodes = ["temperature"]
homie_device_state = None

def publish_homie_version(client):
    topic = "{}/{}/{}".format(homie_prefix, homie_device_id, "$homie")
    client.publish(topic, homie_version, 1, True)

def advertize_device():
    client.publish("homie/homie-3-temperature-sensor/$state", "init", 1, True)
    publish_device_info()
    publish_node_info()
    client.publish("homie/homie-3-temperature-sensor/$state", "ready", 1, True)

def publish_device_info():
    client.publish("homie/homie-3-temperature-sensor/$homie", homie_version, 1, True)
    client.publish("homie/homie-3-temperature-sensor/$name", homie_device_name, 1, True)
    client.publish("homie/homie-3-temperature-sensor/$localip", homie_device_ip, 1, True)
    client.publish("homie/homie-3-temperature-sensor/$mac", homie_device_mac, 1, True)
    client.publish("homie/homie-3-temperature-sensor/$fw/name", "sensor-firmware", 1, True)
    client.publish("homie/homie-3-temperature-sensor/$fw/version", "1.0.0", 1, True)
    client.publish("homie/homie-3-temperature-sensor/$nodes", "temperature", 1, True)
    client.publish("homie/homie-3-temperature-sensor/$implementation", "homie-python", 1, True)
    client.publish("homie/homie-3-temperature-sensor/$stats/interval", "60", 1, True)

def publish_node_info():
    client.publish("homie/homie-3-temperature-sensor/temperature/$name", "Indoor temperature", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/$type", "Temperature", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/$properties", "temperature", 1, True)

    client.publish("homie/homie-3-temperature-sensor/temperature/temperature", 23, 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/temperature/$name", "Indoor temperature", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/temperature/$settable", "false", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/temperature/$unit", "ºC", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/temperature/$datatype", "float", 1, True)
    client.publish("homie/homie-3-temperature-sensor/temperature/temperature/$format", "-20:120", 1, True)

def on_connect(client, userdata, flags, rc):
    print("Connected with result code {}".format(rc))
    advertize_device()


def on_message(client, userdata, msg):
    print("{} -> {}".format(msg.topic, msg.payload))


def loop():
    last_temperature_time = 0
    while True:
        if (time.time() - last_temperature_time) > 5:
            temperature_value = random.uniform(20,25)
            print("Sending tempertaure value: {0:.2f}ºC".format(temperature_value))
            client.publish("homie/homie-3-temperature-sensor/temperature/temperature", "{0:.2f}".format(temperature_value), 1, True)
            last_temperature_time = time.time()


client = mqtt.Client(mqtt_client_id)
client.on_connect = on_connect
client.on_message = on_message

client.will_set("homie/homie-3-temperature-sensor/$state", "lost", 1, True)

client.connect(mqtt_broker_host, mqtt_broker_port, mqtt_client_keepalive)

client.loop_start()

try:
    loop()
except KeyboardInterrupt:
    print("Stopping client")
    client.publish("homie/homie-3-temperature-sensor/$state", "disconnected", 1, True)
    client.disconnect()
    client.loop_stop()
```

Anyway, the first "Offline" state of the Thing keeps happening and IMHO that should not happen, the binding should correctly detect which is the broker where the Thing was discovered. Also, in the properties of the Thing the "homieversion" is not defined, the field is empty.

Thanks again and sorry for the mistake :cry: 

Best regards,

Aitor

**Comment by davidgraeff** [comment]
I know about some start up hickups and other cosmetic problems. That is why there is a fix already in progress: #6339
I very much appreciate that you like the support.

**Comment by davidgraeff** [comment]
@htreu Close please

**Comment by bodiroga** [comment]
Hi @davidgraeff!

I don't know what I'm doing wrong, but I'm still seeing that the Homie Thing is set to offline (with the "Bridge Offline" message) after importing it from the inbox. If I edit the Thing and set the bridge manually, everything works as expected.

What I have done (I have repeated the process twice with the same result):
- Stop openHAB2 (sudo systemctl stop openhab2.service)
- Uninstall and purge my openHAB2 snapshot installation (sudo apt remove --purge openhab2)
- Stop the mosquitto server (I will use the embedded broker) (sudo systemctl stop mosquitto.service)
- Install openHAB2 again (sudo apt install openhab2 -> Build #1412, it should contain the fixes)
- Install the Embedded MQTT Broker addon (PaperUI -> Add-ons -> Misc - Embedded MQTT Broker)
- Install the new MQTT Binding (PaperUI -> Add-ons -> Bindings -> MQTT Binding)
- Import the Embedded broker connection as a Thing (it was autodiscovered in the Inbox)
- Execute the python script that simulates a Homie V3 device (I have also tried with a real NodeMCU with a custom implementation of homie-esp8266, PR in a few hours)
- Import the "Homie Device" device (the device appeared automatically in the inbox, no need to make a manual scan)
- Check the Thing status -> "OFFLINE - BRIDGE_OFFLINE" and in the edit page, I can see that the bridge has not been automatically selected. :cry:
- If a set the bridge by hand, everything works :+1: 

Sorry for being such a pain in the ass, but for me the problem has not been solved.

Thanks for your incredible work once again and best regards,

Aitor

**Comment by bodiroga** [comment]
Ouch, I think that I may have spoken too soon. It seems, from the Karaf console, that this is the MQTT binding version shipped with openHAB2 Build #1412:

198 ? Active   ?  80 ? 0.10.0.201810290804    ? Eclipse SmartHome Embedded MQTT Broker
199 ? Active   ?  80 ? 0.10.0.201810290804    ? Eclipse SmartHome MQTT Transport Bundle
200 ? Active   ?  80 ? 0.10.0.201810290804    ? Eclipse SmartHome MQTT Binding
201 ? Active   ?  80 ? 0.10.0.201810290804    ? Eclipse SmartHome MQTT Thing Binding

And having a look at the PR section of this repository, I can confirm that the #6339 PR was merged on "30 oct. 2018 16:04 CET". One day later :man_facepalming:

**Comment by bodiroga** [comment]
Great, the new version sets the bridge perfectly, sorry for the last couple of messages :+1:
